realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -864,21 +867,216 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
+    if (isDifferentVersions) {
+        LangBindHelper::advance_read(*SG(bgSharedGroupPtr), *CH(replicationPtr));
+    }
+
+    return query;
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindWithHandover(
+    JNIEnv* env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr, jlong fromTableRow)
+{
+    TR_ENTER()
+    try {
+        std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, replicationPtr, queryPtr);
+        Table* table = query->get_table().get();
+
+        if (!QUERY_VALID(env, query.get())) {
+            return 0;
+        }
+
+        // It's valid to go 1 past the end index
+        if ((fromTableRow < 0) || (S(fromTableRow) > table->size())) {
+            // below check will fail with appropriate exception
+            (void) ROW_INDEX_VALID(env, table, fromTableRow);
+            return 0;
+        }
+
+        size_t r = query->find(S(fromTableRow));
+        if (r == not_found) {
+            return 0;
+        } else {
+            // handover the result
+            Row row = (*table)[r];
+            std::unique_ptr <SharedGroup::Handover<Row>> handover = SG(
+                    bgSharedGroupPtr)->export_for_handover(row);
+            return reinterpret_cast<jlong>(handover.release());
+        }
+
+    } CATCH_STD()
+    return 0;
+}
+
+
 JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAll(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlong start, jlong end, jlong limit)
 {
-    Query* pQuery = Q(nativeQueryPtr);
-    Table* pTable = pQuery->get_table().get();
-    if (!QUERY_VALID(env, pQuery) ||
-        !ROW_INDEXES_VALID(env, pTable, start, end, limit))
+    TR_ENTER()
+    Query* query = Q(nativeQueryPtr);
+    Table* table = query->get_table().get();
+    if (!QUERY_VALID(env, query) ||
+        !ROW_INDEXES_VALID(env, table, start, end, limit))
         return -1;
     try {
-        TableView* pResultView = new TableView( pQuery->find_all(S(start), S(end), S(limit)) );
-        return reinterpret_cast<jlong>(pResultView);
+        TableView* tableView = new TableView( query->find_all(S(start), S(end), S(limit)) );
+        return reinterpret_cast<jlong>(tableView);
     } CATCH_STD()
     return -1;
 }
 
+JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllWithHandover
+  (JNIEnv *env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr, jlong start, jlong end, jlong limit)
+  {
+      TR_ENTER()
+      try {
+          std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, replicationPtr, queryPtr);
+          Table* table = query->get_table().get();
+
+          if (!QUERY_VALID(env, query.get()) ||
+              !ROW_INDEXES_VALID(env, table, start, end, limit)) {
+              return 0;
+          }
+
+          // run the query
+          TableView* tableView = new TableView(query->find_all(S(start), S(end), S(limit)));
+
```

Comment:
Release the tableView in the error cases to avoid leaks. Or simply use unique_ptr? if `export_for_handover` failed tableView will be leaked. Should the `tableView` be released even in the normal case?